modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,209 @@
+
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+
+  params.addParam<MultiMooseEnum>("constraint_types",
+                                  HomogenizationConstants::mooseConstraintType,
+                                  "Type of each constraint: strain or stress");
+
+  return params;
+}
+
+TotalLagrangianStressDivergence::TotalLagrangianStressDivergence(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Kernel>(parameters),
+    _ld(getParam<bool>("large_kinematics")),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_nums(_ndisp),
+    _disp_vars(_ndisp),
+    _pk1(getMaterialPropertyByName<RankTwoTensor>("pk1_stress")),
+    _dpk1(getMaterialPropertyByName<RankFourTensor>("pk1_jacobian")),
+    _macro_gradient_num(isCoupledScalar("macro_gradient", 0) ? coupledScalar("macro_gradient") : 0),
+    _macro_gradient(isCoupledScalar("macro_gradient", 0) ? getScalarVar("macro_gradient", 0)
+                                                         : nullptr),
+    _indices(HomogenizationConstants::indices.at(_ld)[_ndisp - 1])
+{
+  // Setup variables
+  for (unsigned int i = 0; i < _ndisp; i++)
+  {
+    _disp_nums[i] = coupled("displacements", i);
+    _disp_vars[i] = getVar("displacements", i);
+  }
+
+  if (isCoupledScalar("macro_gradient", 0))
+  {
+    // Check the order of the scalar variable
+    unsigned int needed = HomogenizationConstants::required.at(_ld)[_ndisp - 1];
+    if (_macro_gradient->order() != needed)
+      mooseError("The homogenization macro gradient variable must have order ", needed);
+
+    // Check the number of constraints
+    auto types = getParam<MultiMooseEnum>("constraint_types");
+    if (types.size() != needed)
+      mooseError("The kernel must be supplied ", needed, " constraint types.");
+
+    // Setup the types of constraints
+    if (isCoupledScalar("macro_gradient", 0))
+    {
+      auto types = getParam<MultiMooseEnum>("constraint_types");
+      for (unsigned int i = 0; i < types.size(); i++)
+        _ctypes.push_back(static_cast<HomogenizationConstants::ConstraintType>(types.get(i)));
+    }
+  }
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpResidual()
+{
+  return _pk1[_qp].row(_component) * _grad_test[_i][_qp];
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpJacobian()
+{
+  Real value = 0.0;
+  value += materialJacobian(_component, _component, _grad_test[_i][_qp], _grad_phi[_j][_qp]);
+
+  return value;
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  Real value = 0.0;
+
+  for (unsigned int cc = 0; cc < _ndisp; cc++)
+  {
+    if (jvar == _disp_nums[cc])
+    {
+      value +=
+          materialJacobian(_component, cc, _grad_test[_i][_qp], _disp_vars[cc]->gradPhi()[_j][_qp]);
+      break;
+    }
+  }
+
+  return value;
+}
+
+Real
+TotalLagrangianStressDivergence::materialJacobian(unsigned int i,
```

Comment:
I use a one liner now (though it's different than the `elastiJacobian` method because stabilization introduces "non i" terms to the trial function gradients).